File: balance.py
import time
import schedule
import requests
import base64
import hmac
import hashlib
from supabase import create_client
import telebot

# Kết nối Supabase
url = "https://tramnanrzruzvkehpydl.supabase.co"
key = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InRyYW1uYW5yenJ1enZrZWhweWRsIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NDU3NTM1NTMsImV4cCI6MjA2MTMyOTU1M30.L0Ytkxi80AbYjkjpDfGyQtfyfqjfHLF98OrVce9Hi-0"
supabase = create_client(url, key)

# Cài đặt bot Telegram
telegram_token = '7648930428:AAFDIISTuWwa-aNmyWgItakI_tMwuTEXNkw'  # Thay bằng token bot Telegram của bạn
bot = telebot.TeleBot(telegram_token)


# ==== Hàm lấy thông tin API từ Supabase ====
def get_api_keys_from_supabase(user_id):
    try:
        response = supabase.table("users").select("apikey_okx", "secret_okx", "pass_okx", "apikey_binance", "secret_binance", "pass_binance").eq("user_id", user_id).execute()
        if response.data:
            return response.data[0]
        else:
            print("Không tìm thấy người dùng với user_id:", user_id)
            return None
    except Exception as e:
        print(f"Đã xảy ra lỗi khi lấy thông tin từ Supabase: {e}")
        return None

# ==== Hàm tạo chữ ký HMAC SHA256 (dùng cho OKX) ====
def sign(message, secret):
    return base64.b64encode(
        hmac.new(secret.encode(), message.encode(), hashlib.sha256).digest()
    ).decode()

# ==== Hàm lấy thời gian từ server OKX ====
def get_timestamp():
    try:
        r = requests.get('https://www.okx.com/api/v5/public/time')
        ts = r.json()['data'][0]['ts']
        return str(int(ts) / 1000)
    except Exception as e:
        print("Đã xảy ra lỗi khi lấy thời gian từ server OKX:", e)
        return None

# ==== Hàm lấy số dư OKX ====
def get_okx_balance(user_id):
    print("Bắt đầu lấy số dư từ OKX...")

    # Lấy API Key, Secret từ Supabase
    user_data = get_api_keys_from_supabase(user_id)
    if not user_data:
        return None

    api_key = user_data['apikey_okx']
    api_secret = user_data['secret_okx']
    passphrase = user_data['pass_okx']

    timestamp = get_timestamp()
    if not timestamp:
        print("Không thể lấy thời gian từ server OKX.")
        return None

    method = "GET"
    request_path = "/api/v5/account/balance"
    body = ""
    prehash_string = f"{timestamp}{method}{request_path}{body}"
    signature = sign(prehash_string, api_secret)
    headers = {
        'OK-ACCESS-KEY': api_key,
        'OK-ACCESS-SIGN': signature,
        'OK-ACCESS-TIMESTAMP': timestamp,
        'OK-ACCESS-PASSPHRASE': passphrase,
        'Content-Type': 'application/json'
    }
    
    response = requests.get("https://www.okx.com/api/v5/account/balance", headers=headers)
    if response.status_code == 200:
        data = response.json()
        balances = data['data'][0]['details']
        usdt_balance = None
        for b in balances:
            if b['ccy'] == 'USDT':
                usdt_balance = b['availBal']
                break
        return usdt_balance
    else:
        print("Lỗi khi lấy thông tin số dư từ OKX:", response.json())
        return None

# ==== Hàm lấy số dư Binance ====
def get_binance_balance(user_id):
    print("Bắt đầu lấy số dư từ Binance...")

    # Lấy API Key, Secret từ Supabase
    user_data = get_api_keys_from_supabase(user_id)
    if not user_data:
        return None

    api_key = user_data['apikey_binance']
    api_secret = user_data['secret_binance']

    # Khởi tạo client Binance
    client = Client(api_key, api_secret)

    try:
        # Lấy số dư tài khoản
        account_info = client.get_account()
        balances = account_info['balances']
        
        # Tìm USDT trong danh sách số dư
        usdt_balance = None
        for balance in balances:
            if balance['asset'] == 'USDT':
                usdt_balance = balance['free']
                break
        return usdt_balance
    except Exception as e:
        print(f"Đã xảy ra lỗi khi lấy số dư từ Binance: {e}")
        return None

# ==== Hàm lấy số dư cho cả OKX và Binance ====
def get_balance_for_user(user_id):
    print(f"Đang lấy số dư cho người dùng {user_id}...")

    # Lấy số dư từ OKX và Binance
    okx_balance = get_okx_balance(user_id)
    binance_balance = get_binance_balance(user_id)

    if okx_balance is not None and binance_balance is not None:
        # Gửi thông tin số dư qua Telegram
        message = f"Số dư của bạn:\n- OKX USDT: {okx_balance}\n- Binance USDT: {binance_balance}"
        bot.send_message(user_id, message)
    else:
        bot.send_message(user_id, "Không thể lấy số dư từ các sàn giao dịch.")

# ==== Lệnh Telegram xử lý khởi tạo cho người dùng ====
@bot.message_handler(commands=['start'])
def handle_start(message):
    user_id = message.chat.id  # Lấy user_id từ Telegram
    print(f"Đã nhận user_id: {user_id} từ Telegram")

    # Gọi hàm lấy số dư từ cả OKX và Binance
    get_balance_for_user(user_id)

    # Trả lời tin nhắn chào mừng
    bot.reply_to(message, "Chào bạn! Đang lấy số dư tài khoản của bạn từ OKX và Binance...")

# ==== Cài đặt lịch để cập nhật mỗi 30 giây ====
schedule.every(30).seconds.do(lambda: get_balance_for_user('user_abc'))  # Thay 'user_abc' bằng user_id thực tế

# ==== Chạy vòng lặp để kiểm tra và thực hiện các tác vụ đã lên lịch ====
while True:
    schedule.run_pending()
    time.sleep(1)

File: balance.js
const fetch = require('node-fetch');
const crypto = require('crypto');
const { Telegraf } = require('telegraf');

const supabaseUrl = 'https://tramnanrzruzvkehpydl.supabase.co';
const supabaseKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InRyYW1uYW5yenJ1enZrZWhweWRsIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NDU3NTM1NTMsImV4cCI6MjA2MTMyOTU1M30.L0Ytkxi80AbYjkjpDfGyQtfyfqjfHLF98OrVce9Hi-0'; // để nguyên key gốc của bạn
const bot = new Telegraf('7648930428:AAFDIISTuWwa-aNmyWgItakI_tMwuTEXNkw');

const userIds = new Set();

// ==== Lấy API Key từ Supabase ====
async function getApiKeys(userId) {
  const res = await fetch(`${supabaseUrl}/rest/v1/users?user_id=eq.${userId}`, {
    headers: {
      apikey: supabaseKey,
      Authorization: `Bearer ${supabaseKey}`,
      'Content-Type': 'application/json',
    },
  });
  const data = await res.json();
  return data[0];
}

// ==== Cập nhật số dư USDT vào Supabase ====
async function updateBalancesInSupabase(userId, usdtOkx, usdtBinance) {
  try {
    await fetch(`${supabaseUrl}/rest/v1/users?user_id=eq.${userId}`, {
      method: 'PATCH',
      headers: {
        apikey: supabaseKey,
        Authorization: `Bearer ${supabaseKey}`,
        'Content-Type': 'application/json',
        Prefer: 'return=representation',
      },
      body: JSON.stringify({
        usdt_okx: usdtOkx,
        usdt_binance: usdtBinance,
      }),
    });
  } catch (e) {
    console.error('Lỗi cập nhật Supabase:', e);
  }
}

// ==== Lấy số dư từ OKX ====
async function getOkxBalance(userData) {
  try {
    const timestamp = new Date().toISOString();

    const prehash = timestamp + 'GET' + '/api/v5/account/balance';
    const signature = crypto
      .createHmac('sha256', userData.secret_okx)
      .update(prehash)
      .digest('base64');

    const res = await fetch('https://www.okx.com/api/v5/account/balance', {
      method: 'GET',
      headers: {
        'OK-ACCESS-KEY': userData.apikey_okx,
        'OK-ACCESS-SIGN': signature,
        'OK-ACCESS-TIMESTAMP': timestamp,
        'OK-ACCESS-PASSPHRASE': userData.pass_okx,
        'Content-Type': 'application/json',
      },
    });

    const json = await res.json();
    if (json.code !== '0') {
      console.error('OKX API error:', json.msg);
      return null;
    }

    const balances = json.data[0].details;
    const usdt = balances.find((b) => b.ccy === 'USDT');
    return usdt?.availBal || null;
  } catch (e) {
    console.error('OKX Error:', e);
    return null;
  }
}

// ==== Lấy số dư từ Binance ====
async function getBinanceBalance(userData) {
  try {
    const timeRes = await fetch('https://api.binance.com/api/v3/time');
    const timeData = await timeRes.json();
    const timestamp = timeData.serverTime;

    const query = `timestamp=${timestamp}`;
    const signature = crypto
      .createHmac('sha256', userData.secret_binance)
      .update(query)
      .digest('hex');

    const res = await fetch(`https://api.binance.com/api/v3/account?${query}&signature=${signature}`, {
      headers: {
        'X-MBX-APIKEY': userData.apikey_binance,
      },
    });

    const json = await res.json();

    if (json.code) {
      console.error('Binance API error:', json.msg || json);
      return null;
    }

    const usdt = json.balances.find((b) => b.asset === 'USDT');
    return usdt?.free || null;
  } catch (e) {
    console.error('Binance Error:', e);
    return null;
  }
}

// ==== Lấy số dư cho user và cập nhật Supabase ====
async function getBalanceForUser(userId) {
  const userData = await getApiKeys(userId);
  if (!userData) {
    console.log(`Không tìm thấy user ${userId}`);
    return;
  }

  const [okx, binance] = await Promise.all([
    getOkxBalance(userData),
    getBinanceBalance(userData),
  ]);

  // Cập nhật số dư vào Supabase
  await updateBalancesInSupabase(userId, okx, binance);

  let message = 'Không thể lấy số dư từ các sàn.';
  if (okx && binance) {
    message = `Số dư của bạn:\n- OKX USDT: ${okx}\n- Binance USDT: ${binance}`;
  }

  await bot.telegram.sendMessage(userId, message);
}

// ==== Xử lý /start từ Telegram ====
bot.start((ctx) => {
  const userId = ctx.message.chat.id;
  userIds.add(userId);
  ctx.reply('Chào bạn! Đang lấy số dư từ OKX và Binance...');
  getBalanceForUser(userId);
});

// ==== Chạy định kỳ mỗi 30 giây ====
setInterval(() => {
  for (const userId of userIds) {
    getBalanceForUser(userId);
  }
}, 30000);

// ==== Khởi động bot ====
bot.launch();
